# Kaggle ML Challenge 2 #

**Team Members:**
    Diksha, 
    Roshan

**Introducttion:**

The Getting Started Notebook is always a great help for me to get started with the Kaggle challenge. Starting with NMF and understanding the NMF pseudocode was quite challenging at first so we tried going with Multivariate Regression to see how the algorithm works and performs.

Multivariate Regression can also be termed as a supervised machine learning technique. So, this explains that our model is used on the train rna and adt dataset through which we understand the relationship between them and predict the test_adt


**Importing Dataset:**

Firstly, we read the training and test csv file using the read_csv() function as below.

In [25]:
import numpy as np
import pandas as pd

In [26]:
test_rna = pd.read_csv('test_set_rna.csv')
test_rna.head()

Unnamed: 0  test_sample_1  test_sample_2  test_sample_3  test_sample_4  \
0      RPL22       2.500262       2.900439       2.953532       2.999922   
1      PARK7       0.000000       1.351622       0.000000       1.572109   
2       ENO1       2.877991       1.351622       1.185308       0.000000   
3       RBP7       1.885984       0.000000       0.000000       0.000000   
4     AGTRAP       1.885984       1.351622       0.000000       0.000000   

   test_sample_5  test_sample_6  test_sample_7  test_sample_8  test_sample_9  \
0       3.181827       3.267155       2.899259       3.157958       2.942404   
1       0.997105       1.424210       0.000000       1.705567       0.000000   
2       0.617999       0.000000       0.000000       0.000000       1.702991   
3       0.000000       0.000000       0.000000       0.000000       2.942404   
4       0.000000       0.000000       0.000000       0.000000       0.000000   

   ...  test_sample_991  test_sample_992  test_sample_993  test_sample_994  \
0  ...         1.879607         1.560418         3.692494         2.745812   
1  ...         1.328524         0.000000         0.000000         0.000000   
2  ...         0.000000         0.000000         1.389076         1.767994   
3  ...         0.000000         0.000000         0.000000         0.000000   
4  ...         0.000000         0.000000         0.000000         2.371941   

   test_sample_995  test_sample_996  test_sample_997  test_sample_998  \
0         3.154371         3.595094         2.925132         3.116989   
1         0.000000         0.000000         1.928458         1.969240   
2         1.555977         0.000000         1.593159         0.000000   
3         0.000000         0.000000         1.085065         0.000000   
4         0.000000         0.000000         1.593159         0.000000   

   test_sample_999  test_sample_1000  
0         4.084811          3.827927  
1         0.000000          0.000000  
2         0.000000          0.000000  
3         0.000000          0.000000  
4         0.000000          0.000000  

[5 rows x 1001 columns]

In [27]:
training_adt = pd.read_csv('training_set_adt.csv')
training_adt.head()

Unnamed: 0  training_sample_1  training_sample_2  training_sample_3  \
0        CD11a           0.413866           0.294449           0.531469   
1        CD11c           1.299088           0.136347           1.043159   
2        CD123           0.474643           0.205216           0.520798   
3  CD127-IL7Ra           0.402731           0.757762           0.354215   
4         CD14           0.527878           0.160274           0.298374   

   training_sample_4  training_sample_5  training_sample_6  training_sample_7  \
0           0.263649           0.438042           0.686346           0.730096   
1           0.246849           0.284185           0.404868           0.603104   
2           0.073185           0.265222           0.564916           0.426254   
3           0.872654           1.337662           1.864847           0.615251   
4           0.189462           0.272238           0.419696           0.587582   

   training_sample_8  training_sample_9  ...  training_sample_3991  \
0           0.299949           1.111089  ...              0.526016   
1           0.354919           0.482943  ...              0.512543   
2           0.426254           0.321830  ...              0.607170   
3           0.984558           0.303224  ...              1.124305   
4           0.323845           0.272238  ...              0.396581   

   training_sample_3992  training_sample_3993  training_sample_3994  \
0              1.348110              0.512804              0.408960   
1              0.490426              0.371848              0.320178   
2              0.426254              0.375404              0.724171   
3              0.320511              1.764897              0.303224   
4              0.396581              0.527878              0.323845   

   training_sample_3995  training_sample_3996  training_sample_3997  \
0              1.166204              0.713007              0.368821   
1              2.240665              0.609743              0.217903   
2              1.967158              0.426254              0.073185   
3              0.386819              1.796728              1.191072   
4              2.448333              0.606716              0.160274   

   training_sample_3998  training_sample_3999  training_sample_4000  
0              1.064956              1.060479              0.808950  
1              0.880499              0.475405              0.371848  
2              1.714842              0.686671              0.265222  
3              0.249493              2.544464              1.850105  
4              1.459272              0.625492              0.507158  

[5 rows x 4001 columns]

In [28]:
training_rna = pd.read_csv("training_set_rna.csv")
training_rna.head()

Unnamed: 0  training_sample_1  training_sample_2  training_sample_3  \
0      RPL22           1.946768           2.993597           2.327345   
1      PARK7           1.387045           0.000000           1.727283   
2       ENO1           0.000000           1.747344           0.000000   
3       RBP7           0.000000           0.000000           0.000000   
4     AGTRAP           0.000000           0.000000           0.000000   

   training_sample_4  training_sample_5  training_sample_6  training_sample_7  \
0           3.355145           2.992697           3.409415           3.241927   
1           0.000000           0.000000           0.000000           1.013754   
2           0.000000           0.000000           1.297154           1.013754   
3           0.000000           0.000000           0.000000           0.000000   
4           0.000000           0.000000           0.000000           1.013754   

   training_sample_8  training_sample_9  ...  training_sample_3991  \
0           3.219029           2.797789  ...              2.791541   
1           0.000000           1.814190  ...              1.808623   
2           1.757990           1.814190  ...              0.000000   
3           0.000000           0.000000  ...              0.000000   
4           0.000000           0.000000  ...              0.000000   

   training_sample_3992  training_sample_3993  training_sample_3994  \
0              2.979010              3.776064              1.593003   
1              1.977239              0.000000              0.000000   
2              0.000000              1.660966              0.000000   
3              0.000000              0.000000              0.000000   
4              0.000000              0.000000              0.000000   

   training_sample_3995  training_sample_3996  training_sample_3997  \
0              3.770663              2.535107              3.322802   
1              0.000000              1.362115              0.000000   
2              0.000000              0.000000              0.000000   
3              0.000000              0.000000              0.000000   
4              0.000000              0.000000              0.000000   

   training_sample_3998  training_sample_3999  training_sample_4000  
0              2.018552              3.790125              3.059502  
1              0.000000              1.151552              0.000000  
2              2.018552              0.000000              1.622036  
3              0.000000              0.000000              0.000000  
4              0.000000              0.000000              0.000000  

[5 rows x 4001 columns]

I then proceeded with dropping the first column from all the dataset to proceed with the matrix multiplication.

In [29]:
update_test_rna = test_rna.drop(test_rna.columns[0], axis =1)
update_test_rna.head()

test_sample_1  test_sample_2  test_sample_3  test_sample_4  test_sample_5  \
0       2.500262       2.900439       2.953532       2.999922       3.181827   
1       0.000000       1.351622       0.000000       1.572109       0.997105   
2       2.877991       1.351622       1.185308       0.000000       0.617999   
3       1.885984       0.000000       0.000000       0.000000       0.000000   
4       1.885984       1.351622       0.000000       0.000000       0.000000   

   test_sample_6  test_sample_7  test_sample_8  test_sample_9  test_sample_10  \
0       3.267155       2.899259       3.157958       2.942404        2.670896   
1       1.424210       0.000000       1.705567       0.000000        1.701890   
2       0.000000       0.000000       0.000000       1.702991        2.299441   
3       0.000000       0.000000       0.000000       2.942404        0.000000   
4       0.000000       0.000000       0.000000       0.000000        0.000000   

   ...  test_sample_991  test_sample_992  test_sample_993  test_sample_994  \
0  ...         1.879607         1.560418         3.692494         2.745812   
1  ...         1.328524         0.000000         0.000000         0.000000   
2  ...         0.000000         0.000000         1.389076         1.767994   
3  ...         0.000000         0.000000         0.000000         0.000000   
4  ...         0.000000         0.000000         0.000000         2.371941   

   test_sample_995  test_sample_996  test_sample_997  test_sample_998  \
0         3.154371         3.595094         2.925132         3.116989   
1         0.000000         0.000000         1.928458         1.969240   
2         1.555977         0.000000         1.593159         0.000000   
3         0.000000         0.000000         1.085065         0.000000   
4         0.000000         0.000000         1.593159         0.000000   

   test_sample_999  test_sample_1000  
0         4.084811          3.827927  
1         0.000000          0.000000  
2         0.000000          0.000000  
3         0.000000          0.000000  
4         0.000000          0.000000  

[5 rows x 1000 columns]

In [30]:
update_training_adt = training_adt.drop(training_adt.columns[0], axis =1)
update_training_adt.head()

training_sample_1  training_sample_2  training_sample_3  training_sample_4  \
0           0.413866           0.294449           0.531469           0.263649   
1           1.299088           0.136347           1.043159           0.246849   
2           0.474643           0.205216           0.520798           0.073185   
3           0.402731           0.757762           0.354215           0.872654   
4           0.527878           0.160274           0.298374           0.189462   

   training_sample_5  training_sample_6  training_sample_7  training_sample_8  \
0           0.438042           0.686346           0.730096           0.299949   
1           0.284185           0.404868           0.603104           0.354919   
2           0.265222           0.564916           0.426254           0.426254   
3           1.337662           1.864847           0.615251           0.984558   
4           0.272238           0.419696           0.587582           0.323845   

   training_sample_9  training_sample_10  ...  training_sample_3991  \
0           1.111089            0.272143  ...              0.526016   
1           0.482943            0.198129  ...              0.512543   
2           0.321830            0.265222  ...              0.607170   
3           0.303224            0.911364  ...              1.124305   
4           0.272238            0.160274  ...              0.396581   

   training_sample_3992  training_sample_3993  training_sample_3994  \
0              1.348110              0.512804              0.408960   
1              0.490426              0.371848              0.320178   
2              0.426254              0.375404              0.724171   
3              0.320511              1.764897              0.303224   
4              0.396581              0.527878              0.323845   

   training_sample_3995  training_sample_3996  training_sample_3997  \
0              1.166204              0.713007              0.368821   
1              2.240665              0.609743              0.217903   
2              1.967158              0.426254              0.073185   
3              0.386819              1.796728              1.191072   
4              2.448333              0.606716              0.160274   

   training_sample_3998  training_sample_3999  training_sample_4000  
0              1.064956              1.060479              0.808950  
1              0.880499              0.475405              0.371848  
2              1.714842              0.686671              0.265222  
3              0.249493              2.544464              1.850105  
4              1.459272              0.625492              0.507158  

[5 rows x 4000 columns]

In [31]:
update_training_rna = training_rna.drop(training_rna.columns[0], axis = 1)
update_training_rna.head()

training_sample_1  training_sample_2  training_sample_3  training_sample_4  \
0           1.946768           2.993597           2.327345           3.355145   
1           1.387045           0.000000           1.727283           0.000000   
2           0.000000           1.747344           0.000000           0.000000   
3           0.000000           0.000000           0.000000           0.000000   
4           0.000000           0.000000           0.000000           0.000000   

   training_sample_5  training_sample_6  training_sample_7  training_sample_8  \
0           2.992697           3.409415           3.241927           3.219029   
1           0.000000           0.000000           1.013754           0.000000   
2           0.000000           1.297154           1.013754           1.757990   
3           0.000000           0.000000           0.000000           0.000000   
4           0.000000           0.000000           1.013754           0.000000   

   training_sample_9  training_sample_10  ...  training_sample_3991  \
0           2.797789            3.629944  ...              2.791541   
1           1.814190            0.000000  ...              1.808623   
2           1.814190            0.000000  ...              0.000000   
3           0.000000            0.000000  ...              0.000000   
4           0.000000            0.000000  ...              0.000000   

   training_sample_3992  training_sample_3993  training_sample_3994  \
0              2.979010              3.776064              1.593003   
1              1.977239              0.000000              0.000000   
2              0.000000              1.660966              0.000000   
3              0.000000              0.000000              0.000000   
4              0.000000              0.000000              0.000000   

   training_sample_3995  training_sample_3996  training_sample_3997  \
0              3.770663              2.535107              3.322802   
1              0.000000              1.362115              0.000000   
2              0.000000              0.000000              0.000000   
3              0.000000              0.000000              0.000000   
4              0.000000              0.000000              0.000000   

   training_sample_3998  training_sample_3999  training_sample_4000  
0              2.018552              3.790125              3.059502  
1              0.000000              1.151552              0.000000  
2              2.018552              0.000000              1.622036  
3              0.000000              0.000000              0.000000  
4              0.000000              0.000000              0.000000  

[5 rows x 4000 columns]

In [32]:
# Converting the dataframe into array
testrna_array = update_test_rna.values
testrna_array.shape

(639, 1000)

In [33]:
trainingadt_array = update_training_adt.values
trainingadt_array.shape

(25, 4000)

In [34]:
trainingrna_array = update_training_rna.values
trainingrna_array.shape

(639, 4000)

**Multivariate Regression**

We use the multivariate regression technique, Ax= y to model the relationship between the train_rna and train_adt. We used gradient descent optimization algorithm to find x and once we had obtained x, we mapped x into the new dataset test_rna to predict the cooresponding values of test_adt.


In [35]:
# setting A as train_
A = trainingrna_array
A.shape

(639, 4000)

In [36]:
#setting y as train_adt
y = trainingadt_array
y.shape

(25, 4000)

In [37]:
# Converting it in the form of ax=b through transposing and matrix multiplication
a = np.dot(A,A.T)
a.shape


(639, 639)

In [38]:
b = np.dot(A,y.T)
b.shape

(639, 25)

We then used gradient descent to find the x. **Gradient Descent** is an iterative optimization algorithm used to find the minimum of the function.

Our gd_solve function takes the parameter a, b which we get from our above code. The gd_solve also takes the parameter max_iter which states the maximum iteration, and tol_cutoff states the tolerance level at which the algorithm will stop.

Our function first sets x as array of zeros with the same dimension as b. Our x is updated through running in the loop until the maximum iteration is achieved. As our loop starts, it will execute 639 times going for each row of the array and will calculate the ith row of b dividing by the ith row of the diagonal a which then updates the solution x.

In [39]:
# Applying gradient descent algorithm to find the value of x

def gd_solve(a, b, max_iter=100, tol_cutoff=1e-15):
    x = np.zeros_like(b)
    for iter in range(1, max_iter+1):
        tol = np.zeros_like(b)
#         print(tol.shape)
        #print(len(b))
        for i in range(len(b)):
            #print(b[i])
            #print(a[i,i])
            z = b[i] / a[i, i]
            x[i] += z
            b -= z * a[:, i].reshape((len(b), 1))
            tol += np.abs(z / x[i])
        if np.all(tol < tol_cutoff):
            print('Break')
            break
    return x



x2 = gd_solve(a, b)

In [40]:
x2.shape

(639, 25)

We tried to see if our above function worked for predicting the test_adt and followed the below steps to predict the test_adt using our test_rna dataset. After saving the submisson folder, we then changed on converting our submission file to Kaggle submission format in the R from the code provided by the Professor.

In [41]:
#Transposing the test_rna and finding the dot product of test_rna and x2
ytest = np.dot(testrna_array.T,x2)
ytest.shape

(1000, 25)

In [43]:
ytest1=ytest.T
ytest1.shape

(25, 1000)

In [44]:
np.savetxt('submission.csv', ytest1, delimiter=',')

From our first Kaggle submission, our Pearson Correlation Coefficient score was **0.80178** which was above the good score from the evaluation. So, we knew that we were on right track and started tweaking with our Multivariate Regression code to see how it can perform better. We firstly started with changing the maximum iteration and tolerance cutoff which increased our score to **0.80217**. In the meantime, we both started working on understanding the NMF algorithm and started researching and trying on both of our ends.

We also tried Gaussian Elimination to see how it works, but our score was relatively same performing the Gaussian Elimination. After researching more, we came across about learning rate and tried implementing learning rate in our Multivariate Regression code.

**Learning Rate** is a hyperparameter that controls the step size at which the model's parameters are updated during training. So, through learning rate, we know how much parameters of the model are adjusted in response to the estimated error at each iteration of the training.

In [45]:
def gd_solve(a, b, learning_rate = 0.001, max_iter=100, tol_cutoff=1e-15):
    x = np.zeros_like(b)
    for iter in range(1, max_iter+1):
        tol = np.zeros_like(b)
        for i in range(len(b)):
            #print(b[i])
            z = b[i] / a[i, i]
            x[i] += learning_rate * z
            b -= learning_rate * z * a[:, i].reshape((len(b), 1))
            tol += np.abs(z / x[i])
        if np.all(tol < tol_cutoff):
            print('Break')
            break
    return x



x2 = gd_solve(a, b, learning_rate = 0.0001)

In [46]:
x2.shape

(639, 25)

Implementing the learning rate in our gradient descent algorithm, our score increased to **0.80235**. We then started tweaking with the learning rate to find our optimal learning rate and we found our optimal learning rate was at 0.0001 which increased our score to **0.81240**. We also tried implementing regularization but it was not suitable for our algorithm. So, we then started tweaking with maximum iteration and tolerance cutoff to see if we could achieve the cutoff point. After many try, we found our maximum iteration as 18000 which provided a high score of **0.81440**.

In [69]:
#Transposing the test_rna and finding the dot product of test_rna and x2
ytest = np.dot(testrna_array.T,x2)
ytest.shape

(1000, 25)

In [70]:
ytest1=ytest.T
ytest1.shape

(25, 1000)

In [71]:
np.savetxt('submission.csv', ytest1, delimiter=',')

After downloading our submission.csv file, we then referred back to R studio to convert the submission file into Kaggle submission format by reshaping them from wide to tall following below code in R.


Convert to Kaggle submission format by reshaping from wide to tall:
```{r}
test <- as.matrix(read.csv("submission.csv", row.names = 1))
```



```{R}
sample_submission <- reshape2::melt(test)
head(sample_submission)
```

```{R}
sample_submission <- data.frame(
  "ID" = paste0("ID_", 1:nrow(sample_submission)), 
  "Expected" = sample_submission$value)
head(sample_submission)
```

Write the final submission file.

```{R}
write.csv(sample_submission, "my_submission.csv", row.names = FALSE)
```

**Other Approach Implementation:**

**NMF Implementation**

Non-negative Matrix Factorization is one of the dimension reduction technique that is used to find the approximate value as of the original matrix when we multiply the W and H matrices.

We then use gradient descent first to find out the value for x. The gradient descent for the Multivariate Regression and NMF is different as it includes non-negativity in the gradient descent. The non-negativity constraint can lead the factorization to be more interpretable.

In [53]:
#Function that performs gradient descent 

def gd_solve(a, b, max_iter=100, tol_cutoff=1e-8):
    x = np.zeros(len(b))
    for iter in range(1, max_iter+1):
        tol = 0
        for i in range(len(b)):
            if a[i, i] == 0:
                z = 0
            else:
                z = b[i] / a[i, i]
            
            if x[i] + z < 0:
                z = -x[i]
                x[i]=0
            else:    
                x[i] = x[i] + z
            
            b = b - z * a[:, i]
            tol = tol + abs(z / (x[i] + 1e-15))
        if tol < tol_cutoff:
            break
    return x

We then worked on creating the function for the NMF from the pseudocode. Firstly, W is randomly intialized and we set H as all zero. First on each iteration, the function updates H from A and ranomly intialized W by soliving the linear system of equations. The NMF function uses the gd_solve function(gradient descent algorithm) to solve it. Once, we have found H, the function then updates W matrix using the linear system of equation. To calculate the W matrix, it also uses the gd_solve(gradient descent) algorithm. 

In [65]:
# Function that performs NMF

def nmf(A, k, max_iter=10):
    # Initialize W randomly and set H as 0
    m, n = A.shape
    W = abs(np.random.rand(m, k))
    H = np.zeros((k, n))
    at=A.T

    # Iteratively update W and H
    for i in range(max_iter):
        # Update H
        a = np.dot(W.T, W)
        #print('a',a.shape)
        for j in range(n):
            
            b = np.dot(W.T, A[:, j])
            x= gd_solve(a,b)
            H[:,j] = x
       
        
        # Update W
        aw = np.dot(H,H.T)
        #print('aw is: ',aw.shape)
        for i in range(0,m):
            bw = np.dot(H, at[:,i])
            #print('bw is: ',bw.shape)
            x1= gd_solve(aw,bw)
            W.T[:, i] = x1
        
 
    return W, H

In [66]:
k = 3
w1,h1=nmf(A,k)

In [67]:
w1
w1.shape

(639, 3)

In [68]:
h1
h1.shape

(3, 4000)

We first tried to see if our NMF algorithm was implemented and tried to see if we faced any errors. We were mostly focused on implementing the pseudocode, so we didn't focus on scaling W and H due to the time limitation as we were also working on tweaking the Multivariate Regression. Firstly, we tried working with k =2 to see how our NMF model was accurate trying to visualize it using the scatterplot. Also, we tried implementing k as 3 and 15 where 15 provided more accurate model.

**Conclusion**

We tried working with Multivariate Regression which provided a good score, however, we also wanted to try NMF to predict and see how it performs and compare both the models. Unfortunately, with limited time, we couldn't complete implementing NMF model as we were yet to work on scaling the W and H matrix then mask the train_adt to predict it. We also want to work on implementing NMF during our Spring Break to see how it performs and compare both the models.


**Reference**:
1. https://www.mygreatlearning.com/blog/understanding-learning-rate-in-machine-learning/#:~:text=Learning%20rate%20is%20used%20to,function%20is%20minimized%20or%20not.
2. https://blog.acolyer.org/2019/02/18/the-why-and-how-of-nonnegative-matrix-factorization/#:~:text=Nonnegative%20matrix%20factorization%20(NMF)%20has,Lee%20and%20Seung%20in%201999.
3. https://www.kaggle.com/code/zdebruine/getting-started-with-ml-challenge-2-prediction
